source/common/filter/config_discovery_impl.h

```diff
@@ -0,0 +1,292 @@
+#pragma once
+
+#include "envoy/config/core/v3/extension.pb.h"
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/config/subscription.h"
+#include "envoy/filter/config_provider_manager.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/config/subscription_base.h"
+#include "common/config/utility.h"
+#include "common/init/manager_impl.h"
+#include "common/init/target_impl.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Filter {
+
+class FilterConfigSubscription;
+
+using FilterConfigSubscriptionSharedPtr = std::shared_ptr<FilterConfigSubscription>;
+
+/**
+ * Implementation of a filter config provider using discovery subscriptions.
+ **/
+class DynamicFilterConfigProviderImplBase : public Config::ExtensionConfigProviderBase {
+public:
+  DynamicFilterConfigProviderImplBase(FilterConfigSubscriptionSharedPtr&& subscription,
+                                      const std::set<std::string>& require_type_urls);
+  virtual ~DynamicFilterConfigProviderImplBase();
+  const Init::Target& initTarget() const { return init_target_; }
+
+  // Config::ExtensionConfigProviderBase
+  const std::string& name() override;
+  void validateConfig(const ProtobufWkt::Any& proto_config) override;
+
+private:
+  FilterConfigSubscriptionSharedPtr subscription_;
+  const std::set<std::string> require_type_urls_;
+
+  // Local initialization target to ensure that the subscription starts in
+  // case no warming is requested by any other filter config provider.
+  Init::TargetImpl init_target_;
+};
+
+/**
+ * Base class for a provider managing subscriptions.
+ **/
+class FilterConfigProviderManagerImplBase {
+public:
+  virtual ~FilterConfigProviderManagerImplBase() = default;
+
+protected:
+  FilterConfigSubscriptionSharedPtr
+  getSubscription(const envoy::config::core::v3::ConfigSource& config_source,
+                  const std::string& name, Server::Configuration::FactoryContext& factory_context,
+                  const std::string& stat_prefix);
+  virtual void validateTypeUrl(const std::string& filter_config_name, absl::string_view type_url,
+                               bool is_terminal) const PURE;
+  std::set<std::string>
+  validateConfigSource(const std::string& filter_config_name,
+                       const envoy::config::core::v3::ExtensionConfigSource& config_source,
+                       bool is_terminal) const;
+
+private:
+  absl::flat_hash_map<std::string, std::weak_ptr<FilterConfigSubscription>> subscriptions_;
```

Comment:
Prefer manage provider to manage subscription to address https://github.com/envoyproxy/envoy/issues/14934. see https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/common/router/rds_impl.h#L260